File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    private StringBuilder path = new StringBuilder();
    private List<String> ans = new ArrayList<>();

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return ans;
        }

        String[] digitsMap = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        backtracking(digits, digitsMap, 0);
        return ans;
    }

    private void backtracking(String digits, String[] digitsMap, int startIdx) {
        if (path.length() == digits.length()) {
            ans.add(path.toString());
            return;
        }

        int digit = digits.charAt(startIdx) - '0'; // get 2
        String letters = digitsMap[digit]; // get "abc"
        for (int i = 0; i < letters.length(); i++) {
            path.append(letters.charAt(i));
            backtracking(digits, digitsMap, startIdx + 1);
            path.deleteCharAt(path.length() - 1);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        digits_map = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        ans = []

        if len(digits) == 0:
            return ans

        def backtracking(start, path) -> None:
            if len(path) == len(digits):
                ans.append(path)
                return
            
            for i in range(start, len(digits)):
                digit = digits[i]
                word = digits_map[ord(digit) - ord('0')]

                for char in word:
                    backtracking(i + 1, path + char)
        
        backtracking(0, "")
        return ans